### PR TITLE
⚠️ Use limited reader in webhooks

### PR DESCRIPTION
**Pull request by inteon**
This PR adds a LimitedReader which prevents DOS attacks due to OOM caused by the `io.ReadAll` function.

Depends on https://github.com/kubernetes-sigs/controller-runtime/pull/2604.

Code is based on https://github.com/kubernetes/kubernetes/blob/c28c2009181fcc44c5f6b47e10e62dacf53e4da0/staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go

### CODE REVIEW

**Comment by alvaroaleman** [comment]
@inteon would you mind splitting this up in one PR per change, so discussions can stay targeted and the changelog will have a more meaningful entry than "Improve Webhook"?

**Comment by inteon** [comment]
> @inteon would you mind splitting this up in one PR per change, so discussions can stay targeted and the changelog will have a more meaningful entry than "Improve Webhook"?

Ok, this PR now focuses on adding `limited reader`.

File: pkg/webhook/admission/http.go [hunk 1]
```diff
@@ -34,6 +34,9 @@ import (
 var admissionScheme = runtime.NewScheme()
 var admissionCodecs = serializer.NewCodecFactory(admissionScheme)
 
+// based on https://github.com/kubernetes/kubernetes/blob/c28c2009181fcc44c5f6b47e10e62dacf53e4da0/staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go
+var maxRequestSize = int64(3 * 1024 * 1024)
```
Comment (alvaroaleman) [inline pkg/webhook/admission/http.go:38]: This is technically a breaking change. More importantly, how do we know this size is appropriate? I guess for admission we can inferr that a request contains at most two objects which each have a maximum size of one meg plus the request itself which is also not likely to exceed one meg. But how about requests in the authentication webhook?

Comment (inteon) [reply pkg/webhook/admission/http.go:38]: Upstream, this value seems to come from here: https://github.com/kubernetes/apiserver/blob/master/pkg/server/config.go#L433-L442C5

Comment (inteon) [reply pkg/webhook/admission/http.go:38]: So, k/k has made a mistake using that value for this admission webhook?: https://github.com/kubernetes/kubernetes/blob/c28c2009181fcc44c5f6b47e10e62dacf53e4da0/staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go

Comment (inteon) [reply pkg/webhook/admission/http.go]: @alvaroaleman I updated the maxRequestSize based on the reasoning you explained and added comments: https://github.com/kubernetes-sigs/controller-runtime/pull/2598/commits/9dcfc53b6dd7852c484034b763b015258dd382a6

File: pkg/webhook/admission/http.go [hunk 2]
```diff
@@ -55,12 +73,19 @@ func (wh *Webhook) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	body, err := io.ReadAll(limitedReader)
 	if err != nil {
 		wh.getLogger(nil).Error(err, "unable to read the body from the incoming request")
 		wh.writeResponse(w, Errored(http.StatusBadRequest, err))
 		return
 	}
+	if limitedReader.N <= 0 {
+		err := fmt.Errorf("request entity is too large; limit is %d bytes", maxRequestSize)
+		wh.getLogger(nil).Error(err, "unable to read the body from the incoming request; limit reached")
+		wh.writeResponse(w, Errored(http.StatusBadRequest, err))
```
Comment (sbueringer) [inline pkg/webhook/admission/http.go:86]:
```suggestion
		wh.writeResponse(w, Errored(http.StatusRequestEntityTooLarge, err))
```

Seems more precise

File: pkg/webhook/admission/http.go [hunk 3]
```diff
@@ -34,6 +34,24 @@ import (
 var admissionScheme = runtime.NewScheme()
 var admissionCodecs = serializer.NewCodecFactory(admissionScheme)
 
+// adapted from https://github.com/kubernetes/kubernetes/blob/c28c2009181fcc44c5f6b47e10e62dacf53e4da0/staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go
+//
+// From https://github.com/kubernetes/apiserver/blob/d6876a0600de06fef75968c4641c64d7da499f25/pkg/server/config.go#L433-L442C5:
+//
+//	     1.5MB is the recommended client request size in byte
+//		 the etcd server should accept. See
```
Comment (sbueringer) [inline pkg/webhook/admission/http.go:42]: Given all the stuff people are doing today. Can we really still assume that controller-runtime is only used in environments where etcd is used?

File: pkg/webhook/authentication/http.go [hunk 4]
```diff
@@ -34,6 +34,20 @@ import (
...
+// adapted from https://github.com/kubernetes/kubernetes/blob/c28c2009181fcc44c5f6b47e10e62dacf53e4da0/staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go
+//
+// From https://github.com/kubernetes/apiserver/blob/d6876a0600de06fef75968c4641c64d7da499f25/pkg/server/config.go#L433-L442C5:
+//
+//	     1.5MB is the recommended client request size in byte
+//		 the etcd server should accept. See
+//		 https://github.com/etcd-io/etcd/blob/release-3.4/embed/config.go#L56.
+//		 A request body might be encoded in json, and is converted to
+//		 proto when persisted in etcd, so we allow 2x as the largest request
+//		 body size to be accepted and decoded in a write request.
```
Comment (sbueringer) [inline pkg/webhook/authentication/http.go:46]: Q: What does this have to do with the TokenReview request?

File: pkg/webhook/admission/http.go (hunk 3 above)
Comment (vincepri) [reply pkg/webhook/admission/http.go:42]: Probably not, but this limiter should only apply to admission webhooks; which seems fair?

File: pkg/webhook/authentication/http.go (hunk 4 above)
Comment (inteon) [reply pkg/webhook/authentication/http.go:46]: You are right, the TokenReview request does not come from etcd.
Not sure what a better upper-limit would be.
I did read somewhere that JWTs have a upper size limit because they have to be includable in the header of a HTTP request, maybe that is a starting point?

File: pkg/webhook/admission/http.go (hunk 3 above)
Comment (sbueringer) [reply pkg/webhook/admission/http.go:42]: I assume the apiserver calls admission webhooks independent of what datastore is used behind it.

I wasn't thinking about non-apiserver scenarios. More about the cases where folks are just not using etcd as storage

Comment (sbueringer) [reply pkg/webhook/admission/http.go:42]: Nevermind. I missed that every object that is stored goes through the apiserver which enforces the limit

File: pkg/webhook/authentication/http.go (hunk 4 above)
Comment (sbueringer) [reply pkg/webhook/authentication/http.go:46]: After a quick look at google/stackoverflow it seems that neither JWT nor HTTP defines a max size for a JWT / a HTTP header. Just that there are limits that webservers have in place: https://stackoverflow.com/questions/686217/maximum-on-http-header-values

It seems usually HTTP header limits are between 4 and 48k. Maybe let's use 1 MB as a limit which should be low enough given what we're protecting against.

Then I would add a comment that the TokenReview mostly contains a bearer token which at most should have a few kb's of size, so we picked 1 MB to have plenty of buffer.

Comment (sbueringer) [reply pkg/webhook/authentication/http.go:46]: @alvaroaleman @vincepri ^^ any other ideas on how to pick a limit?

Comment (vincepri) [reply pkg/webhook/authentication/http.go:46]: The logic above makes sense to me, in both cases I'd add a comment that if this happens, folks should open an issue, and link to https://github.com/kubernetes-sigs/controller-runtime/issues/new

**Comment by sbueringer** [comment]
@inteon PTAL when you have some time :)

**Comment by inteon** [comment]
@sbueringer I updated the limits and comments based on the review feedback.
Please let me know in case I missed something.

**Comment by sbueringer** [comment]
Thx!

/lgtm

/assign @alvaroaleman @vincepri

File: pkg/webhook/admission/http.go [hunk 5]
```diff
@@ -34,6 +34,26 @@ import (
...
+//		 body size to be accepted and decoded in a write request.
+//
+// For the admission request, we can infer that it contains at most two objects
+// (the old and new versions of the object being admitted), each of which can
+// be at most 3MB in size. For the rest of the request, we can assume that
+// it will be less than 1MB in size. Therefore, we can set the max request
+// size to 7MB.
+// If your use case requires larger max request sizes, please
+// open an issue (https://github.com/kubernetes-sigs/controller-runtime/issues/new).
+var maxRequestSize = int64(7 * 1024 * 1024)
```
Comment (alvaroaleman) [inline pkg/webhook/admission/http.go:55]: `const`

File: pkg/webhook/authentication/http.go [hunk 6]
```diff
@@ -34,6 +34,13 @@ import (
 var authenticationScheme = runtime.NewScheme()
 var authenticationCodecs = serializer.NewCodecFactory(authenticationScheme)
 
+// The TokenReview resource mostly contains a bearer token which
+// at most should have a few KB's of size, so we picked 1 MB to
+// have plenty of buffer.
+// If your use case requires larger max request sizes, please
+// open an issue (https://github.com/kubernetes-sigs/controller-runtime/issues/new).
+var maxRequestSize = int64(1 * 1024 * 1024)
```
Comment (alvaroaleman) [inline pkg/webhook/authentication/http.go:42]: const

**Review by alvaroaleman — APPROVED** [review]
